core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -722,6 +728,45 @@ object Validator {
     }
   }
 
+  private def validateDirectives(types: List[__Type]): IO[ValidationError, Unit] = {
+
+    def validateArguments(args: Map[String, InputValue], errorContext: String): IO[ValidationError, Unit] = {
+      val explanatoryText =
+        s"""The directive argument must not have a name which begins with the characters "__" (two underscores)"""
+      val argumentErrorContext: String => String = argName => s"Argument '$argName' of $errorContext"
+      IO.foreach_(args.map(_._1))(argName =>
+        doesNotStartWithUnderscore[String](argName, identity, argumentErrorContext(argName), explanatoryText)
+      )
+    }
+
+    def validateDirective(directive: Directive, errorContext: String) = {
+      val directiveErrorContext = s"Directive '${directive.name}' of $errorContext"
+      for {
+        _ <- doesNotStartWithUnderscore(directive, directiveErrorContext)
+        _ <- validateArguments(directive.arguments, directiveErrorContext)
+      } yield ()
+    }
+
+    def validateDirectives[T](
+      t: T,
+      directivesExtractor: T => Option[List[Directive]],
+      errorContext: String
+    ): IO[ValidationError, Unit] =
+      IO.foreach_(directivesExtractor(t).getOrElse(List.empty))(validateDirective(_, errorContext))
+
+    IO.foreach_(types) { t =>
+      val typeErrorContext                    = s"Type '${t.name.getOrElse("")}'"
+      val fieldErrorContext: String => String = fieldName => s"Field '$fieldName' of $typeErrorContext"
+      for {
+        _ <- validateDirectives[__Type](t, _.directives, typeErrorContext)
+        _ <- IO.foreach_(t.fields(__DeprecatedArgs(Some(true))).getOrElse(List.empty[__Field]))(f =>
```

Comment:
There can also be directives in `__Type.inputFields` and `__Field.args`: both contain `__InputValue` which has `directives`.

Reply:
I haven't seen those, added!